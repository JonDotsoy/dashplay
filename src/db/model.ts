import { Connection } from "../interfaces/connection";
import { Mutate, Query, Schema, mutate, query } from "./query";

export class Model<S = any> {
  private constructor(
    readonly targetName: string,
    readonly schema: Schema<S>,
    readonly connection: Connection,
  ) { }

  async init() {
    await this.connection.defineSchema(this.targetName, this.schema)
  }

  static async create<S = any>(
    targetName: string,
    schema: Schema<S>,
    connection: Connection) {
    const model = new Model(
      targetName,
      schema,
      connection
    );

    await model.init();

    return model;
  }

  query(): Query<S> {
    return query<S>(this.targetName)
  }

  mutate(): Mutate<S> {
    return mutate<S>(this.targetName)
  }

  async runQuery(exp: Query<S>): Promise<S[] | null> {
    return await this.connection.runQuery(exp)
  }

  async runMutate(exp: Mutate<S>) {
    return await this.connection.runMutate(exp)
  }
}
