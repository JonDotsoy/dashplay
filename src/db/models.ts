import { AppContext } from "../app/app-context";
import { Connection } from "../interfaces/connection";
import { ConnectionFactory } from "./connection-factory";
import { Model } from "./model";
import { Schema, SchemaToType, SchemaTypes, ObjFromSchema } from "./query";


export class Models {
  private constructor(
    readonly ctx: AppContext,
    readonly logGroups: Model<ObjFromSchema<typeof LogGroupSchema>>,
  ) { }

  static async create(ctx: AppContext<any>) {
    return new Models(
      ctx,
      await Model.create(
        'log-groups',
        LogGroupSchema,
        ctx.connection,
      ),
    )
  }
}

export type LogStreamManifest = ObjFromSchema<typeof LogStreamManifestSchema>

export const EventMessageSchema = Schema.createSchema({
  id: SchemaTypes.string,
  message: SchemaTypes.string,
  timeStamp: SchemaTypes.number,
})

export const LogGroupSchema = Schema.createSchema({
  id: SchemaTypes.string,
  name: SchemaTypes.string,
  createdAt: SchemaTypes.number,
}, {
  indexes: ['name']
})

export type EventMessage = ObjFromSchema<typeof EventMessageSchema>

export const LogStreamManifestSchema = Schema.createSchema({
  id: SchemaTypes.string,
  name: SchemaTypes.string,
  timestampFrom: SchemaTypes.number,
  timestampTo: SchemaTypes.number,
}, {
  indexes: ['name']
})
