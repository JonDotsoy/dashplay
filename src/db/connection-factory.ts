import { Engine, type Connection, type InitConnectionOptions, type ConnectionConstructor } from "../interfaces/connection";
import { SQLiteConnection } from "./connections/sqlite/sqlite-connection";

export class ConnectionFactory {
  private constructor(readonly connection: Connection) { }

  static connectionEnginesList: Record<Engine, ConnectionConstructor> = {
    ['sqlite']: SQLiteConnection
  }

  static async create<E extends string>(reference: string, options: InitConnectionOptions<E>) {
    const engine = options.engine

    if (!Reflect.has(this.connectionEnginesList, engine)) throw new Error(`Invalid engine ${engine}`)

    const Engine = Reflect.get(this.connectionEnginesList, engine)

    const connection: Connection = new Engine(reference, options);
    await connection.init();
    return connection;
  }
}
