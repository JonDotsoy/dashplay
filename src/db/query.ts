export enum SchemaTypes {
  string,
  number,
  boolean,
  /** @deprecated */
  date,
}

export type TransformValue<T> = T extends SchemaTypes.string ? string
  : T extends SchemaTypes.number ? number
  : T extends SchemaTypes.boolean ? boolean
  : T extends SchemaTypes.date ? Date
  : never;

export type TransformSchemaTypes<T> = T extends Record<string, SchemaTypes>
  ? { [k in keyof T]: TransformValue<T[k]> }
  : never;

export type ObjFromSchema<T> = T extends Schema<infer R>
  ? R
  : never;

export interface SchemaOptions<T = any> {
  indexes: KeysOfSchema<T>[]
}

export class Schema<T = any> {
  constructor(readonly def: Record<string, SchemaTypes>, readonly options?: SchemaOptions<T>) { }

  static createSchema<T extends Record<string, SchemaTypes>>(
    def: T,
    options?: SchemaOptions
  ) {
    return new Schema<TransformSchemaTypes<T>>(def, options);
  }
}

export type SchemaToType<T> = T extends Schema<infer R> ? R : any

type KeysOfSchema<T> = T extends Schema<infer O> ? keyof O : any;
type TypesOfSchemaByKey<T, K> = T extends Schema<infer O>
  ? O extends Record<any, any> ? O[K] : any
  : any;

type PartialObjectBySchema<T> = T extends Schema<infer O> ? Partial<O> : any;

export class Assertion { }

export class EqualTo<S, T extends KeysOfSchema<S>> extends Assertion {
  constructor(readonly field: T, readonly value: TypesOfSchemaByKey<S, T>) {
    super();
  }
}

export class NotEqualTo<S, T extends KeysOfSchema<S>> extends Assertion {
  constructor(readonly field: T, readonly value: TypesOfSchemaByKey<S, T>) {
    super();
  }
}

export class LessThan<S, T extends KeysOfSchema<S>> extends Assertion {
  constructor(readonly field: T, readonly value: TypesOfSchemaByKey<S, T>) {
    super();
  }
}

export class GreaterThan<S, T extends KeysOfSchema<S>> extends Assertion {
  constructor(readonly field: T, readonly value: TypesOfSchemaByKey<S, T>) {
    super();
  }
}

export class LessThanOrEqualTo<S, T extends KeysOfSchema<S>> extends Assertion {
  constructor(readonly field: T, readonly value: TypesOfSchemaByKey<S, T>) {
    super();
  }
}

export class GreaterThanOrEqualTo<S, T extends KeysOfSchema<S>>
  extends Assertion {
  constructor(readonly field: T, readonly value: TypesOfSchemaByKey<S, T>) {
    super();
  }
}

export class Like<S, T extends KeysOfSchema<S>> extends Assertion {
  constructor(readonly field: T, readonly value: string) {
    super();
  }
}

export class Between<S, T extends KeysOfSchema<S>> extends Assertion {
  constructor(
    readonly field: T,
    readonly valueLeft: TypesOfSchemaByKey<S, T>,
    readonly valueRight: TypesOfSchemaByKey<S, T>,
  ) {
    super();
  }
}

export class In<S, T extends KeysOfSchema<S>> extends Assertion {
  constructor(readonly field: T, readonly values: TypesOfSchemaByKey<S, T>[]) {
    super();
  }
}

export class And<S> extends Assertion {
  constructor(readonly assertions: Assertion[]) {
    super();
  }
}

export class Or<S> extends Assertion {
  constructor(readonly assertions: Assertion[]) {
    super();
  }
}

export class Not<S> extends Assertion {
  constructor(readonly andAssertion: And<S>) {
    super();
  }
}

export class Where<S = any> {
  constructor(private query: { addAssertion: (assertion: Assertion) => any }) { }

  equalTo<T extends KeysOfSchema<S>>(
    field: T,
    value: TypesOfSchemaByKey<S, T>,
  ) {
    this.query.addAssertion(new EqualTo(field, value));
    return this;
  }

  notEqualTo<T extends KeysOfSchema<S>>(
    field: T,
    value: TypesOfSchemaByKey<S, T>,
  ) {
    this.query.addAssertion(new NotEqualTo(field, value));
    return this;
  }

  lessThan<T extends KeysOfSchema<S>>(
    field: T,
    value: TypesOfSchemaByKey<S, T>,
  ) {
    this.query.addAssertion(new LessThan(field, value));
    return this;
  }


  greaterThan<T extends KeysOfSchema<S>>(field: T, value: TypesOfSchemaByKey<S, T>) {
    this.query.addAssertion(new GreaterThan<S, T>(field, value))
    return this
  }

  lessThanOrEqualTo<T extends KeysOfSchema<S>>(field: T, value: TypesOfSchemaByKey<S, T>) {
    this.query.addAssertion(new LessThanOrEqualTo<S, T>(field, value))
    return this
  }

  greaterThanOrEqualTo<T extends KeysOfSchema<S>>
    (field: T, value: TypesOfSchemaByKey<S, T>) {
    this.query.addAssertion(new GreaterThanOrEqualTo<S, T>(field, value))
    return this
  }

  like<T extends KeysOfSchema<S>>(field: T, value: string) {
    this.query.addAssertion(new Like<S, T>(field, value))
    return this
  }

  between<T extends KeysOfSchema<S>>(
    field: T,
    valueLeft: TypesOfSchemaByKey<S, T>,
    valueRight: TypesOfSchemaByKey<S, T>,
  ) {
    this.query.addAssertion(new Between<S, T>(field, valueLeft, valueRight))
    return this
  }

  in<T extends KeysOfSchema<S>>(field: T, values: TypesOfSchemaByKey<S, T>[]) {
    this.query.addAssertion(new In<S, T>(field, values))
    return this
  }

  and(...cbs: ((where: Where<S>) => void)[]) {
    const assertions: Assertion[] = []
    for (const cb of cbs) {
      cb(new Where({ addAssertion: assertion => assertions.push(assertion) }))
    }
    this.query.addAssertion(new And<S>(assertions))
    return this
  }

  or(...cbs: ((where: Where<S>) => void)[]) {
    const assertions: Assertion[] = []
    for (const cb of cbs) {
      cb(new Where({ addAssertion: assertion => assertions.push(assertion) }))
    }
    this.query.addAssertion(new Or<S>(assertions))
    return this
  }

  not(cb: (where: Where<S>) => void) {
    const assertions: Assertion[] = []
    cb(new Where({ addAssertion: assertion => assertions.push(assertion) }))
    this.query.addAssertion(new Not<S>(
      new And<S>(assertions)
    ))
    return this
  }
}

export class Query<S = any> {
  constructor(
    readonly target: string,
    readonly columnsSet = new Set<KeysOfSchema<S>>(),
    readonly assertionRef: { current?: And<S> } = {},
    readonly setValues = new Set<unknown[]>(),
  ) { }

  fields(...columns: KeysOfSchema<S>[]) {
    for (const column of columns) this.columnsSet.add(column);
    return this;
  }

  set(...values: PartialObjectBySchema<S>[]) {
    this.setValues.add(values);
    return this;
  }

  addAssertion(assertion: Assertion) {
    this.assertionRef.current = this.assertionRef.current ?? new And([])
    this.assertionRef.current.assertions.push(assertion);
    return this;
  }

  where(cb: (where: Where<S>) => void) {
    cb(new Where(this));
    return this;
  }
}

export enum MutateType {
  SET,
  DELETE
}

export class Mutate<S = any> {
  constructor(
    readonly target: string,
    readonly columnsSet = new Set<KeysOfSchema<S>>(),
    readonly assertionRef: { current?: And<S> } = {},
    readonly setValues = new Set<PartialObjectBySchema<S>>(),
    private type: MutateType | null = null
  ) { }

  set(...values: PartialObjectBySchema<S>[]) {
    for (const value of values) {
      for (const key of Object.keys(value)) this.columnsSet.add(key as KeysOfSchema<S>)
      this.setValues.add(value);
    }
    return this;
  }

  addAssertion(assertion: Assertion) {
    this.assertionRef.current = this.assertionRef.current ?? new And([])
    this.assertionRef.current.assertions.push(assertion);
    return this;
  }

  delete(cb: (where: Where<S>) => void) {
    cb(new Where(this));
    return this;
  }

  where(cb: (where: Where<S>) => void) {
    cb(new Where(this));
    return this;
  }
}

export const query = <S = any>(targetName: string) => new Query<S>(targetName)
export const mutate = <S = any>(targetName: string) => new Mutate<S>(targetName)
