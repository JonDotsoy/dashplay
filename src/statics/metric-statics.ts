import ms from "ms"
import { EventMessage } from "../db/models"
import { atom } from "nanostores"
import type { Accumulation, MetricStaticsModule, MetricStaticsOption, ReducerDataPoints, ResultAggregate, TimeStamp } from "./types"


export class MetricStatics {
  interval: number

  constructor(
    ctx: unknown,
    readonly options: MetricStaticsOption,
  ) {
    this.interval = typeof options.interval === 'string' ? ms(options.interval) : options.interval
  }

  getInitialAccumulation(): Accumulation {
    if (typeof this.options.initialAccumulation === 'function') {
      return this.options.initialAccumulation()
    }
    return this.options.initialAccumulation ?? {}
  }

  async *calculateDataPoints(records: AsyncIterable<EventMessage> | Iterable<EventMessage>, options?: { onChangePartialResultAggregate?: (resultAggregate: ResultAggregate, unSubscribe: () => void) => void }): AsyncGenerator<ResultAggregate> {
    const toTimeStampKey = (timestamp: number) => Math.floor(timestamp / this.interval) * this.interval

    let currentResultAggregate = atom<ResultAggregate>({
      timeStamp: NaN,
      value: this.getInitialAccumulation(),
    })

    const onChangePartialResultAggregate = options?.onChangePartialResultAggregate

    if (onChangePartialResultAggregate) {
      let unSubscribe: null | (() => void) = null
      unSubscribe = currentResultAggregate.subscribe((resultAggregate) => {
        onChangePartialResultAggregate(resultAggregate, () => unSubscribe?.())
      })
    }

    for await (const event of records) {
      let resultAggregate = currentResultAggregate.get()
      const timeStampKey = toTimeStampKey(event.timeStamp)

      if (resultAggregate.timeStamp !== timeStampKey) {
        if (!Number.isNaN(resultAggregate.timeStamp)) yield resultAggregate
        resultAggregate = {
          timeStamp: timeStampKey,
          value: this.getInitialAccumulation(),
        }
      }

      resultAggregate.value = this.options.reducer(
        resultAggregate.value,
        event,
      );

      currentResultAggregate.set(resultAggregate);
    }

    const resultAggregate = currentResultAggregate.get()
    if (!Number.isNaN(resultAggregate.timeStamp)) yield resultAggregate
  }

  static async load(moduleRef: MetricStaticsModule | Promise<MetricStaticsModule>) {
    return new MetricStatics({}, await moduleRef)
  }
}