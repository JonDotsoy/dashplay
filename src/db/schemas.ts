import { Schema, SchemaTypes } from "./query";

const asd = Schema.createSchema({
  id: SchemaTypes.string,
});
