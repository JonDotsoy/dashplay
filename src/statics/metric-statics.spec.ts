import { expect, spyOn, test, Mock, mock } from "bun:test"
import { MetricStatics } from "./metric-statics"
import ms from "ms"
import { ObjFromSchema } from "../db/query"
import { EventMessage } from "../db/models"
import { monotonicFactory } from "ulid"

const ulid = monotonicFactory(() => 0)

const makeEvent = (time: number, message: string = `[message]`): EventMessage => {
  return {
    id: ulid(time),
    message: message,
    timestamp: time,
  }
}


const sampleEvents: EventMessage[] = [
  ...Array(3).fill(0).map(() => makeEvent(ms('12s'))),
  ...Array(2).fill(0).map(() => makeEvent(ms('30s'))),
  ...Array(5).fill(0).map(() => makeEvent(ms('30s'))),
  ...Array(1).fill(0).map(() => makeEvent(ms('35s'))),
  ...Array(4).fill(0).map(() => makeEvent(ms('45s'))),
  ...Array(2).fill(0).map(() => makeEvent(ms('49s'))),
  ...Array(3).fill(0).map(() => makeEvent(ms('51s'))),
  ...Array(6).fill(0).map(() => makeEvent(ms('54s'))),
  ...Array(4).fill(0).map(() => makeEvent(ms('58s'))),
  ...Array(7).fill(0).map(() => makeEvent(ms('64s'))),
  ...Array(3).fill(0).map(() => makeEvent(ms('70s'))),
  ...Array(1).fill(0).map(() => makeEvent(ms('71s'))),
  ...Array(1).fill(0).map(() => makeEvent(ms('91s'))),
]


test('should calculate date points', async () => {
  const metricStatics = new MetricStatics(
    {},
    {
      interval: '30s',
      reducer({ count = 0 }) {
        return {
          count: count + 1
        }
      }
    }
  )

  const result = await Array.fromAsync(metricStatics.calculateDataPoints(sampleEvents))
})

test('should watch last changes on result aggregate', async () => {
  const metricStatics = new MetricStatics(
    {},
    {
      interval: '1s',
      reducer({ count = 0 }) {
        return {
          count: count + 1
        }
      }
    }
  )

  const fn = mock((..._: any): any => undefined)

  await Array.fromAsync(metricStatics.calculateDataPoints(sampleEvents, { onChangePartialResultAggregate: fn }))

  expect(fn).toHaveBeenCalled()
})

test('should load a metric statics module', async () => {
  const metricStatics = await MetricStatics.load(await import("./metric-statics-modules/sample1"))

  await Array.fromAsync(metricStatics.calculateDataPoints(sampleEvents))

  expect(metricStatics).toBeInstanceOf(MetricStatics)
})
