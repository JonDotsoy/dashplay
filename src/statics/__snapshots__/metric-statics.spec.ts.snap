// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should to calculate date points 1`] = `
[
  {
    "timeStamp": 0,
    "value": {
      "count": 3,
    },
  },
  {
    "timeStamp": 30000,
    "value": {
      "count": 27,
    },
  },
  {
    "timeStamp": 60000,
    "value": {
      "count": 11,
    },
  },
  {
    "timeStamp": 90000,
    "value": {
      "count": 1,
    },
  },
]
`;
